File: c/gtk4/application.c
#include <gtk/gtk.h>

static void on_app_activate(GApplication* self, gpointer data);
static void on_app_startup(GApplication* self, gpointer data);

const static gchar* APP_ID = "io.github.Miqueas.GTK-Examples.C.Gtk4.Application";

gint main(gint argc, gchar** argv) {
  GtkApplication* app = gtk_application_new(APP_ID, 0);

  g_signal_connect(app, "startup", G_CALLBACK(on_app_startup), NULL);
  g_signal_connect(app, "activate", G_CALLBACK(on_app_activate), NULL);

  gint result = g_application_run(G_APPLICATION(app), argc, argv);
  g_object_unref(app);

  return result;
}

static void on_app_activate(GApplication* self, gpointer data) {
  g_print("Hello there!\n");
}

static void on_app_startup(GApplication* self, gpointer data) {
  g_print("Initializing the app... Done!\n");
}

File: lua/gtk4/application.lua
local lgi = require("lgi")
local Gtk = lgi.require("Gtk", "3.0")
local Gio = lgi.Gio

-- GtkApplication: Provides an application interface

local appID = "io.github.Miqueas.GTK-Examples.Lua.Gtk4.Application"
local app = Gtk.Application.new(appID, Gio.ApplicationFlags.FLAGS_NONE)

-- I recommend to check out this links for more information:
-- https://wiki.gnome.org/HowDoI/GtkApplication
-- https://developer.gnome.org/gio/stable/GApplication.html
-- https://developer.gnome.org/gtk3/stable/GtkApplication.html
function app:on_startup()
  -- The ::startup signal is called only once, so it's recommended to create/prepare
  -- all the GUI here and then present it on the ::activate signal
  print("app::startup called")
end

function app:on_activate()
  -- The ::activate signal can be triggered indefinitely, so please try to not do heavy
  -- things here. Check out the doc links above for more details
  print("app::activate called")
end

return app:run(arg)